stargazer.js

```diff
@@ -0,0 +1,67 @@
+import Alt from './'
+import { createStore, datasource } from './utils/decorators'
+import axios from 'axios'
+
+const alt = new Alt()
+
+const StargazerActions = alt.generateActions(
+  'fetchingUsers',
+  'usersReceived',
+  'failed'
+)
+
+const StargazerSource = {
+  fetchUsers: {
+    fetch(state) {
+      const url = `https://api.github.com/repos/${state.user}/${state.repo}/stargazers`
+      return axios({ url }).then(response => response.data)
+    },
+
+    cache(state) {
+      return state.users.length ? state.users : null
+    },
+
+    loading: StargazerActions.fetchingUsers,
+
+    success: StargazerActions.usersReceived,
+
+    error: StargazerActions.failed
+  }
+}
+
+@createStore(alt)
+@datasource(StargazerSource)
+class StargazerStore {
+  constructor() {
+    this.user = 'goatslacker'
+    this.repo = 'alt'
+    this.users = []
+    this.errorMessage = null
+    this.isLoading = false
+
+    this.bindListeners({
+      loading: StargazerActions.fetchingUsers,
+      receivedUsers: StargazerActions.usersReceived,
+      failed: StargazerActions.failed
+    })
+  }
+
+  loading() {
+    console.log('STORE: starting to load users')
+    this.isLoading = true
+  }
+
+  failed(e) {
+    console.log('STORE: Uh oh')
+    this.errorMessage = e.statusText || String(e)
+  }
+
+  receivedUsers(users) {
+    console.log('STORE: got users')
+    this.users = users
+    this.errorMessage = null
+  }
+}
+
+StargazerStore.listen((state) => console.log('CHANGED', state.users.length))
+StargazerStore.fetchUsers()
```

Comment:
I could see a potential use case for forcing an update of the cached data, but I think in most cases it can be solved by the implementation in the cache function.  It takes the state as an argument which allows you to determine whether new data is being queried or not.  You could even store the time of the last fetch and only fetch again after a certain period of time.  I think this is pretty flexible since you define what caching means (if you even choose to have it).

Loading should be true when fetch is called.  Loading is the intermediary state before the data has been fetched. It gives you a hook to give the user feedback that the data is on its way.  If you look at the way this is implemented, it calls the loading function if it is present and then immediately starts trying to fetch data.  Since the fetch is async, the loading function allows us to update some state immediately, which like I said before could give indication that loading is occurring, or simply prepare for the fetched data.

There is no inherent abort in place right now, but like you are saying, that could be implemented in conjunction with this async patter for a pretty sweet user experience.